### PR TITLE
combined all CI improvement

### PR DESCRIPTION
**Pull request by Boegie19**
### What does it do?
Shard
Cheche build.
make lint its own thing.
move build to the start of the project.
removed note 16 on request of derrick
### Why is it needed?

To make CI only take 15 to 20 minutes

### How to test it?

Run the CI

### CODE REVIEW

File: .github/workflows/tests.yml
Comment (innerdvations) [inline .github/workflows/tests.yml:404]: We can't remove 16 until we make an official announcement we've dropped support for it (which I'll make a note to do for 4.15.0 regardless of if we've removed it from the CI and engines yet)

[same comment goes for everywhere]

File: .github/actions/run-build/action.yml
Comment (innerdvations) [inline .github/actions/run-build/action.yml:18]:
```suggestion
      run: yarn nx run-many --target=build --nx-ignore-cycles --skip-nx-cache
      
```

File: .github/workflows/tests.yml
Comment (innerdvations) [inline .github/workflows/tests.yml:38]: I think this is fine, lint should (in theory) be independent of node version.

However, can we make it 20 while we're at it just so we're on the current version?

File: .github/workflows/tests.yml
Comment (innerdvations) [inline .github/workflows/tests.yml:419]: We determined that the increase to using 2 shards will be fine, but I worry 6 could push us over our limit since we do use a significant amount already. Let's wait to see how the other PR goes for a few days before adding this.

**Review by innerdvations — COMMENTED** [review]
Once again, thanks for this, should save a lot of time waiting for tests to pass :)

Let's merge the other one and see how it goes, and then next week we'll merge this one.

File: .github/workflows/tests.yml
Comment (Boegie19) [reply .github/workflows/tests.yml:38]: @innerdvations I would like to but lint (node: 18) is required to run soo then the required to run list needs to change and I am not sure if I can do that.

File: .github/workflows/tests.yml
Comment (Boegie19) [reply .github/workflows/tests.yml:419]: @innerdvations Because of the Caching the CI time is down

File: .github/workflows/tests.yml
Comment (innerdvations) [reply .github/workflows/tests.yml:38]: Do you mean getting it past the current CI? If that's the case I can make sure we update on our end to let it through.

**Comment by Boegie19** [comment]
@innerdvations 
notting duration 68m total 14h 7m 1s        Runtime. https://github.com/strapi/strapi/actions/runs/6470232081

only Sharding duration 38m total 14h 52m 16s    Runtime. https://github.com/strapi/strapi/actions/runs/6469037113

both shard and cache  builds duration  28m 11h 14m 38s  runtime  https://github.com/strapi/strapi/actions/runs/6472242597/usage

3 shards and cache builds duration 23m 19s  11h 5m 57s runtime https://github.com/strapi/strapi/actions/runs/6472811551/usage

4 shards and cache buils duration 17m 50s 11h 33m 21s    runtime https://github.com/strapi/strapi/actions/runs/6473313612|

6 shards and cache buils duration 14 to 16m  12h 6m 55s https://github.com/strapi/strapi/actions/runs/6473580168/usage
wirth node 16 removed duration 14 to 16m 8h 22m 53s https://github.com/strapi/strapi/actions/runs/6479023764

**Comment by innerdvations** [comment]
Ok, I have a few other prioties today but will try to get this out soon!

**Comment by Boegie19** [comment]
Think that 5 times sharding is the best since now we are on the edge so it depends what servers we get how fast it will be.

File: .github/workflows/tests.yml
Comment (Boegie19) [reply .github/workflows/tests.yml:38]: Then I will also remove the unit backend test for 16.

**Comment by Boegie19** [comment]
@innerdvations Any update on the timeline for reviewing this?

File: package.json
Comment (Boegie19) [inline package.json:61]: I am only not sure if this is the right solution

File: .github/workflows/tests.yml
Comment (Marc-Roig) [inline .github/workflows/tests.yml:374]: why removing node16 from the matrix from the mysql action?

**Comment by Marc-Roig** [comment]
cc @innerdvations @joshuaellis would love this to be merged as soon as possible so our tests run quicker <3 could you give it a look?

File: .github/actions/run-build/action.yml
Comment (Boegie19) [reply .github/actions/run-build/action.yml:5]: To my knowledge it is to tell github you can run this on ever runner

File: .github/workflows/tests.yml
Comment (innerdvations) [reply .github/workflows/tests.yml:374]: Yep, should be merged in just a few minutes and then this change will disappear 😆

File: .github/actions/run-build/action.yml
Comment (innerdvations) [inline .github/actions/run-build/action.yml:19]: line 19 has whitespace that shouldn't be there

File: package.json
Comment (innerdvations) [reply package.json:61]: Yeah, unfortunately better options would take a lot more effort and considering this is just for internal tooling I don't think it's worth it to refactor generate-app just for this.

I would go with 'no-build' though, with a dash

**Review by innerdvations — COMMENTED** [review]
Just two extremely minor and annoying comments from me, and then I think we're good to go.

**Comment by innerdvations** [comment]
Aaaaaand there was just a new merge conflict in the last 3 minutes since I reviewed 😆 sorry @Boegie19 , once that's fixed we'll race to merge before it happens again

**Comment by Boegie19** [comment]
@innerdvations fixed the conflicts.

**Review by innerdvations — APPROVED** [review]
Just waiting for CI checks to finish

**Comment by Boegie19** [comment]
apparently CI failed to run since something in the yml file is wrong let me fix.

**Review by innerdvations — APPROVED** [review]

**Comment by Boegie19** [comment]
@innerdvations you would have to removed the required for lint and unit_back before merging.

**Pull request merged**
